File: data/sliding-window/subarrays-with-k-different-integers.cpp
#include <vector>
#include <unordered_map>

using namespace std;

int atMostDistinct(vector<int>& nums, int kDistinct) {
    unordered_map<int, int> count;
    int left = 0, res = 0;
    for (int right = 0; right < nums.size(); ++right) {
        if (count[nums[right]] == 0) {
            kDistinct--;
        }
        count[nums[right]]++;
        
        while (kDistinct < 0) {
            count[nums[left]]--;
            if (count[nums[left]] == 0) {
                kDistinct++;
            }
            left++;
        }
        
        res += right - left + 1;
    }
    return res;
}

int subarraysWithKDistinct(vector<int>& nums, int k) {
    return atMostDistinct(nums, k) - atMostDistinct(nums, k - 1);
}

File: data/sliding-window/subarrays-with-k-different-integers.py
from collections import defaultdict

def subarraysWithKDistinct(nums: list[int], k: int) -> int:
    def atMost(k_distinct):
        count = defaultdict(int)
        left = 0
        res = 0
        for right in range(len(nums)):
            if count[nums[right]] == 0:
                k_distinct -= 1
            count[nums[right]] += 1
            
            while k_distinct < 0:
                count[nums[left]] -= 1
                if count[nums[left]] == 0:
                    k_distinct += 1
                left += 1
                
            res += right - left + 1
        return res

    return atMost(k) - atMost(k - 1)
